### PR TITLE
Fix for introduce local variable in lambda deletes unrelated code

### LINKED ISSUE #40745: Introduce local variable in lambda deletes unrelated code
**Version Used**: VisualStudio.16.Preview/16.5.0-pre.1.0+29521.150

**Steps to Reproduce**:

```C#
Task.Run(() => File.Copy("src", Path.Combine("dir", "file")));
```
(or just a lambda declaration)

Introduce local for `Path.Combine("dir", "file")`

**Expected Behavior**:
Either
```C#
string destFileName = Path.Combine("dir", "file");
Task.Run(() => File.Copy("src", destFileName));
```
or
```C#
Task.Run(() =>
{
    string destFileName = Path.Combine("dir", "file");
    File.Copy("src", destFileName)
});
```

**Actual Behavior**:
```C#
Task.Run(() =>
{
    string destFileName = Path.Combine("dir", "file");
});
```
i.e. `File.Copy` is removed

### PR DESCRIPTION
**Pull request by allisonchou**
Resolves #40745.

### CODE REVIEW

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 1]
```diff
@@ -95,13 +95,35 @@ internal partial class CSharpIntroduceVariableService
...
+            {
+                // We don't need to include the rewritten body if it consists of just the identifier name.
+                // However, in all other cases, the rewritten body should be included.
+                var isRewrittenBodyTrivial = rewrittenBody.WalkDownParentheses().IsKind(SyntaxKind.IdentifierName,
+                    out IdentifierNameSyntax identifier) &&
+                    identifier.Identifier.ValueText == newLocalName.Identifier.ValueText;
+
+                if (isRewrittenBodyTrivial)
+                {
+                    newBody = SyntaxFactory.Block(declarationStatement);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: this portion here (from 106-114) seems like some sort of hack to get some old behavior.  could you clarify overall what hte issue is and what sorts of before/after cases we want to be supporting and why?

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 2]
```diff
@@ -95,13 +95,35 @@ internal partial class CSharpIntroduceVariableService
...
+                {
+                    newBody = SyntaxFactory.Block(
+                        declarationStatement,
+                        SyntaxFactory.ExpressionStatement(rewrittenBody, SyntaxFactory.Token(SyntaxKind.SemicolonToken)));
+                }
+            }
+            else
+            {
+                newBody = SyntaxFactory.Block(declarationStatement, SyntaxFactory.ReturnStatement(rewrittenBody));
+            }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:126]: i basically don't understand most of what's going on here.  specifically why this is correct and doing the right thing for all the cases that might come through here

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs (hunk 1 above)
Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: Previously, if the lambda had a void return type, the introduce local refactoring deleted the original code in the lambda statement and only kept the new declaration statement.
e.g. In this case:
```C#
Task.Run(() => File.Copy("src", [|Path.Combine("dir", "file")|]));
```
The refactoring incorrectly turned it into:
```C#
Task.Run(() =>
{
    string destFileName = Path.Combine("dir", "file");
});
```


I believe the original code was written to aid in cases such as:
```C#
Action<int> goo = x => [|x.ToString()|];
```
which, without the original check, would have ended up like:
```C#
Action<int> goo = x =>
{
    string v = x.ToString();
    v;
};
```

The updated code essentially checks to see if the code being replaced with the new variable is trivial or not. (i.e., is it just the identifier? -> if so, we don't need to include it.)

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs (hunk 2 above)
Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:126]: The non-void return case existed in the previous code for cases like:
```C# 
        Func<int, int> f = x => [|x + 1|];
```

```C#
        Func<int, int> f = x =>
        {
            var v = x + 1;
            return v;
        }
```

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: > The updated code essentially checks to see if the code being replaced with the new variable is trivial or not. (i.e., is it just the identifier? -> if so, we don't need to include it.)

Then what *do* we do for:

```Action<int> goo = x => [|x.ToString()|];```?

What do we generate in that case?

Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: The current functionality generates the following for the case, which I believe is correct:
```C#
        Action<int> goo = x =>
        {
            string v = x.ToString();
        };
```

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: ok, that makes sense (and matches what we do with normal methods).  

i'll have to noodle on this a bit.  something doesn't feel really clear ot me here.  

If anything, have you considered having this code just pass a value into the Rewrite function saying it's a void/non-void context and if the expression is being returned or or not?  It might just be simpler to have the rewrite function try to encode this logic instead of effectively "undoing" it here at this point.

Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: I definitely see where you're coming from regarding passing a value into the Rewrite function. I personally think it may make the code a bit messier since this is a C#-specific issue and the Rewrite function is language-agnostic. 

The language-specific method Rewrite calls into doesn't take in the appropriate parameters, so if we decided to add parameters, they would have to be added to the C# and VB implementations. I'm not sure if this would be considered too much clutter, since VB wouldn't be using the added parameters. Or we could add a new abstract method and the VB implementation could return nothing. Do you have a perspective on what approach you think might be best?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:114]: fair point about the shared Vb/C# logic.

what i really don't like is having some function return data to us that we then have to break apart.  especially with realy janky code like:

```c#
                var isRewrittenBodyTrivial = rewrittenBody.WalkDownParentheses().IsKind(SyntaxKind.IdentifierName,
                    out IdentifierNameSyntax identifier) &&
                    identifier.Identifier.ValueText == newLocalName.Identifier.ValueText;
```

if we need to know if the new body is trivial, why can't that be something that the `Rewrite` function returns out to us as a bool since it must have known this when creating the new body in the first place?

--

Basically, we shouldn't have code that has to re-infer information about what's going on from code deeper in the exact same component.  It indicates a band-aid on a problem after the fact instead of an appropriate amount of work at the source to figure out what's going on.  

So if possible, have `Rewrite` figure things out and return enough real information for the higher up caller to do the right hting.  For VB we can likely just ignore the information (or can we?) if it's not doing special stuff with lambdas.  But for c# we can use that extra information to drive a more suitable experience at this layer.

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 3]
```diff
@@ -91,26 +91,81 @@ internal partial class CSharpIntroduceVariableService
             CancellationToken cancellationToken)
         {
             var oldBody = (ExpressionSyntax)oldLambda.Body;
+            var isEntireLambdaBodySelected = oldBody.WalkDownParentheses().Equals(expression.WalkDownParentheses());
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:94]:
```suggestion
            var isEntireLambdaBodySelected = oldBody.Equals(expression.WalkUpParentheses());
```

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 4]
```diff
@@ -91,26 +91,81 @@ internal partial class CSharpIntroduceVariableService
...
+                //     Task.Run(() =>
+                //     {
+                //         string destFileName = Path.Combine("dir", "file");
+                //         File.Copy("src", destFileName);
+                //     });
+                newBody = newBody.AddStatements(
+                    SyntaxFactory.ExpressionStatement(rewrittenBody, SyntaxFactory.Token(SyntaxKind.SemicolonToken)));
+            }
+            else if (!doesDelegateMethodReturnVoid)
+            {
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:149]:
```suggestion
            if (doesDelegateMethodReturnVoid
            {
                if (!isEntireLambdaBodySelected)
                {
                    // Case 1b: The user didn't select the entire lambda body.
                    // e.g.:
                    //     Task.Run(() => File.Copy("src", [|Path.Combine("dir", "file")|]));
                    //
                    // After refactoring:
                    //     Task.Run(() =>
                    //     {
                    //         string destFileName = Path.Combine("dir", "file");
                    //         File.Copy("src", destFileName);
                    //     });
                    newBody = newBody.AddStatements(
                        SyntaxFactory.ExpressionStatement(rewrittenBody, SyntaxFactory.Token(SyntaxKind.SemicolonToken)));
            }
            else
            {
```

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 5]
```diff
@@ -91,26 +91,81 @@ internal partial class CSharpIntroduceVariableService
...
+                //     {
+                //         var v = x + 1;
+                //         return v;
+                //     };
+                newBody = newBody.AddStatements(
+                    SyntaxFactory.ReturnStatement(rewrittenBody));
+            }
 
             // Add an elastic newline so that the formatter will place this new lambda body across multiple lines.
             newBody = newBody.WithOpenBraceToken(newBody.OpenBraceToken.WithAppendedTrailingTrivia(SyntaxFactory.ElasticCarriageReturnLineFeed))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:165]: so, reading hte above, i think it would just be much simpler as:

```c#
if (doesDelegateMethodReturnVoid )
{
     if (isEntireLambdaBodySelected)
         return SyntaxFactory.Block(declarationStatement);
 
     return ...
}
else 
{
     return ...
}
```

Just have the caller fixup the braces.
}

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs [hunk 6]
```diff
@@ -113,6 +110,94 @@ internal partial class CSharpIntroduceVariableService
...
+                //         Task task = M2();
+                //         await task;
+                //     };
+                var compilation = document.SemanticModel.Compilation;
+                if (oldLambda.AsyncKeyword != default &&
+                    (compilation.TaskType() != null && compilation.TaskType().Equals(delegateType.DelegateInvokeMethod.ReturnType) ||
+                    compilation.ValueTaskOfType() != null && compilation.ValueTaskOfType().Equals(delegateType.DelegateInvokeMethod.ReturnType)))
+                {
+                    return false;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:142]:
```suggestion
                if (oldLambda.AsyncKeyword != default)
                {
                    if ((compilation.TaskType() != null && compilation.TaskType().Equals(delegateType.DelegateInvokeMethod.ReturnType) ||
                    (compilation.ValueTaskOfType() != null && compilation.ValueTaskOfType().Equals(delegateType.DelegateInvokeMethod.ReturnType)))
                {
                    return false;
                }
```

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ICompilationExtensions.cs [hunk 7]
```diff
@@ -112,8 +112,12 @@ public static ImmutableArray<IAssemblySymbol> GetReferencedAssemblySymbols(this
         public static INamedTypeSymbol? TaskOfTType(this Compilation compilation)
             => compilation.GetTypeByMetadataName(typeof(Task<>).FullName!);
 
+        public static INamedTypeSymbol? ValueTaskOfType(this Compilation compilation)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ICompilationExtensions.cs:115]:
```suggestion
        public static INamedTypeSymbol? ValueTaskType(this Compilation compilation)
```

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs (hunk 6 above)
Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs:142]: Can we  assume that the ```compilation.TaskType()``` and ```compilation.ValueTaskType()``` calls will never return null? It seems like they both return nullable types. It also doesn't look like ValueTask existed before C# 7.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs]: > Can we assume that the compilation.TaskType() and compilation.ValueTaskType() calls will never return null? 

No we cannot.  First, there's no guarantee that hte user will be compiling against dlls that have that. Second, during project load these dlls can be missing (because the PS hasn't added them yet), and so these methods will return 'null' :)  

It's a pita, but it does happen, so it's good to write the code to be resilient to that.  We alst test for this.  for example if you look at the UseRange (or maybe UseIndex) tests, we tests for cases where this stuff runs in projects where the extra dlls are not there.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs]: See UseRangeOperationTests.TestWithMissingReference.  This is actually a reasonable idea to test almost all features in this condition as it accurately reflects what might happen during solution load.  No biggie if you do/don't add the test.

**Comment by CyrusNajmabadi** [comment]
Great tests.  LGTM.  :shipit:

File: src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs (hunk 6 above)
Comment (allisonchou) [reply src/Features/CSharp/Portable/IntroduceVariable/CSharpIntroduceVariableService_IntroduceLocal.cs]: Gotcha! Will definitely keep that in mind for the future.

**Comment by allisonchou** [comment]
Thank you!
